doc/pamck.1
.\" This file is part of PAM-Modules -*- nroff -*-
.\" Copyright (C) 2001-2015 Sergey Poznyakoff
.\"
.\" PAM-Modules is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 3, or (at your option)
.\" any later version.
.\"
.\" PAM-Modules is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with PAM-Modules.  If not, see <http://www.gnu.org/licenses/>.
.TH PAMCK 1 "March 31, 2014" "PAM-MODULES" "Pam-Modules User Reference"
.SH NAME
pamck \- test PAM stack
.SH SYNOPSIS
.nh
.na
\fBpamck\fR [\fB\-hv\fR] [\fB\-s\fR \fISERVICE\fR]\
 [\fB\-g\fR \fIGROUP\fR] \fIUSER\fR [\fIPASSWORD\fR]
.ad
.hy
.SH DESCRIPTION
Tests the functionality of \fBPAM\fR stack.
.SH OPTIONS
.TP
.B \-h
Print short help summary and exit.
.TP
.B \-v
Print program version and copyright information and exit.
.TP
\fB\-s\fR \fISERVICE\fR
Select service name to use.  Default group is \fBcheck\fR.
.TP
\fB\-g\fR \fIGROUP\fR
Select \fBPAM\fR management group to check.  Allowed arguments are:
.RS
.TP
.B auth
Authentication group.  Call \fBpam_authenticate\fR.
.TP
.B acct
Account management.  Call \fBpam_acct_mgmt\fR.
.TP
.B open
Session management.  Call \fBpam_open_session\fR.
.TP
.B close
Session management.  Call \fBpam_close_session\fR.
.TP
.B pass
Password management.  Call \fBpam_chauthtok\fR.
.RE
.SH EXIT CODES
.TP
.B 0
Successful termination.
.TP
.B 1
Invalid command line usage.
.TP
.B 2
\fBPAM\fR stack returned error.
.SH EXAMPLES
.nr step 1 1
.IP \n[step].
Test if the user \fBsmith\fR can be authenticated using the
\fBcheck\fR service.  The program will ask for password interactively:
.PP
.EX
pamck smith
.EE
.IP \n+[step].
Same as above, but supply password on the command line:
.PP
.EX
pamck smith guessme
.EE
.IP \n+[step].
Check if the user \fBsmith\fR authenticates using the \fBftp\fR service.
.PP
.EX
pamck -s ftp smith
.EE
.SH NOTE
This manpage is a short description of \fBpamck\fR.  For a detailed
discussion, including examples and usage recommendations, refer to the
\fBPAM-modules Manual\fR available in texinfo format.  If the \fBinfo\fR
reader and the tar documentation are properly installed on your
system, the command
.PP
.RS +4
.B info pam-modules
.RE
.PP
should give you access to the complete manual.
.PP
You can also view the manual using the info mode in
.BR emacs (1),
or find it in various formats online at
.PP
.RS +4
.B http://www.gnu.org.ua/software/pam-modules/manual
.RE
.PP
If any discrepancies occur between this manpage and the
\fBPAM-modules Manual\fR, the later shall be considered the authoritative
source.
.SH "SEE ALSO"
.BR pam.conf (5),
.BR pam.d (5),
.BR pam (8).
.SH AUTHORS
Sergey Poznyakoff <gray@gnu.org>
.SH "BUG REPORTS"
Report bugs to <bug\-pam\-modules@gnu.org.ua>.
.SH COPYRIGHT
Copyright \(co 2001-2014 Sergey Poznyakoff
.br
.na
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
.br
.ad
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.\" Local variables:
.\" eval: (add-hook 'write-file-hooks 'time-stamp)
.\" time-stamp-start: ".TH [A-Z_][A-Z0-9_.\\-]* [0-9] \""
.\" time-stamp-format: "%:B %:d, %:y"
.\" time-stamp-end: "\""
.\" time-stamp-line-limit: 20
.\" end:



NEWS
pam-modules -- history of user-visible changes. 2015-01-30
Copyright (C) 2001, 2004-2005, 2007-2012, 2015 Sergey Poznyakoff
See the end of file for copying conditions.

Please send pam-modules bug reports to <bug-pam-modules@gnu.org.ua>


Version 1.9.90 (Git)

* pam_ldaphome reads LDAP configuration from /etc/ldap.conf

This is in addition to its regular configuration file.

* pam_ldaphome runs inirc-command with user privileges

To run the command with root privileges, the configuration
variable initrc-root must be set to true.

* New pam_ldaphome variable: user-keys-boundary

User key files can contain both keys managed by pam_ldaphome and
added by the user.  These two groups of keys must be separated by
a special comment line, which informs pam_ldaphome that all keys
below it must be retained.

This feature is enabled by the user-keys-boundary configuration
setting.  Its value defines a string which, when used  after a
'#' character, forms the delimiting comment.  E.g. if the
configuration file contains:

  user-keys-boundary :user

then the line '#:user' can be used to delimit ldap-synchronized
and user-specific keys.

Version 1.9, 2014-05-21

* New module pam_groupmember

Tests whether the user is a member of one or more groups.

* pam_ldaphome can invoke an external program

An external program defined with the inirc-command keyword is run
in the newly created user's home directory.  It can be used for
per-user customization of the files copied from the skeleton dir.

The examples directory contains a perl program "usergitconfig", which,
when used as inirc-command, initializes the user's .gitconfig file.

* New auxiliary utilities

** ldappubkey

The `ldappubkey' utility is a simple Perl program which takes user
login name as its argument and produces on the standard output public
ssh keys for that user, each on a separate line.  The program is
designed for use with `openssh' version 6.2p1 or higher.

** usergitconfig

Customizes user's `.gitconfig' file using attributes from his LDAP
entry.  This utility can be used with the initrc-command statement
in pam_ldaphome.conf file.

* Bugfixes


Version 1.8, 2013-07-29

* pam_ldaphome

This module creates the user home directory, if it does not
already exist, and updates his `.ssh/authorized_keys' file with the
keys from the LDAP database.

* pam_umotd

Pam_umotd displays a user-specific message of the day.  The text can
be taken either from a disk file, or read from the standard output of
a program launched for that purpose.  This module is Linux-specific.

* Bugfixes
** pam_fshadow made reentrant



Version 1.7, 2011-04-08

* Allow for use of `CALL proc' in MySQL queries.
* Minor bugfixes in pamck.


Version 1.6, 2009-02-25

* pamck

Pamck is a command line utility for checking PAM authentication and
other management groups.  E.g.:

   pamck -s login smith

attempts to authenticate user `smith' using PAM service name `login'.


Version 1.5, 2009-02-17

* Configure

New command line options:

  --disable-fshadow
  --disable-log
  --disable-regex

Improved autodetection of MySQL and PostgreSQL libraries.

Missing prerequisites for any module cause disabling of that module,
but the configuration process continues.


Version 1.4, 2008-03-20

* pam_mysql and pam_pgsql

** Session management

Session management is implemented for both modules.  Session
management queries are `session-start-query' and `session-stop-query'.

** Variable expansion in configuration file.

Old style of variable expansion has been dropped.  The `$name'
notation is used instead.  To convert your old configuration files,
replace %u with $user, and %p with $password.

** setenv-query

This new query allows to store arbitrary data in PAM environment.


Version 1.3, 2008-03-15

* pam_mysql and pam_pgsql

** Configuration file syntax

Long statements can be split over several lines by placing
'\' character at the end of each line.

** ldap passwords

Both modules understand passwords in LDAP form.  A special
configuration file statement `allow-ldap-pass' is provided to control
this feature.  By default, `allow-ldap-pass yes' is assumed.


Version 1.2, 2008-03-14

* Several fixes in debugging code and pam_mysql, pam_pgsql modules.

* pam_fshadow

By default extended regular expressions are used. 

* pam_regex transform=expr

New command line option `transform' allows to rewrite user names.


Version 1.1, 2007-08-11

* pam_fshadow allows to use virtual domains to specify alternate password
databases.  New options: regex, basic, extended, ignore-case, icase
and revert-index.

* pam_regex: ignore-case can be used as an alias to icase.

* New modules

pam_log                 Log arbitrary data
pam_mysql		Authenticate using a MySQL database
pam_pgsql		Authenticate using a PostgreSQL database


Version 1.0

	Added documentation, improved configuration suite.


Version 0.1

	Initial release. See README for short description.

^L
=========================================================================
Copyright information:

Copyright (C) 2001, 2004-2005, 2007-2015 Sergey Poznyakoff

   Permission is granted to anyone to make or distribute verbatim copies
   of this document as received, in any medium, provided that the
   copyright notice and this permission notice are preserved,
   thus giving the recipient permission to redistribute in turn.

   Permission is granted to distribute modified versions
   of this document, or of portions of it,
   under the above conditions, provided also that they
   carry prominent notices stating who last changed them.

Local variables:
mode: outline
paragraph-separate: "[  ]*$"
eval: (add-hook 'write-file-hooks 'time-stamp)
time-stamp-start: "changes. "
time-stamp-format: "%:y-%02m-%02d"
time-stamp-end: "\n"
end:




pamck/pamck.c
/* This file is part of pam-modules.
   Copyright (C) 2009-2012, 2014-2015 Sergey Poznyakoff
  
   This program is free software; you can redistribute it and/or modify it
   under the terms of the GNU General Public License as published by the
   Free Software Foundation; either version 3 of the License, or (at your
   option) any later version.
  
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
  
   You should have received a copy of the GNU General Public License along
   with this program.  If not, see <http://www.gnu.org/licenses/>. */

#include "pamck.h"
char *program_name;

void
usage()
{
	printf("usage: %s [-hv] [-s service] [-g group] user [password]\n",
	       program_name);
}

void
help()
{
	printf("usage: %s [-hv] [-s service] [-g group] user [password]\n",
	       program_name);
	printf("\nReport bugs to <%s>.\n", PACKAGE_BUGREPORT);
	printf("%s home page: <http://www.gnu.org.ua/software/%s/>.\n",
	       PACKAGE_NAME, PACKAGE);
}

void
version()
{
	printf("%s (%s) %s\n", program_name, PACKAGE, PACKAGE_VERSION);
	fputs ("\
Copyright (C) 2009-2012, 2014 Sergey Poznyakoff\n\
\n\
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.\n\
This is free software: you are free to change and redistribute it.\n\
There is NO WARRANTY, to the extent permitted by law.\n\
\n\
", stdout);
}

void
error(int code, const char *fmt, ...)
{
	va_list ap;
	va_start(ap, fmt);
	fprintf(stderr, "%s: ", program_name);
	vfprintf(stderr, fmt, ap);
	fputc('\n', stderr);
	va_end(ap);
	if (code)
		exit(code);
}


struct grouptab {
	char *name;
	char *funcname;
	int (*pam_fn) (pam_handle_t *pamh, int flags);
} grouptab[] = {
	{ "auth", "pam_authenticate", pam_authenticate },
	{ "acct", "pam_acct_mgmt", pam_acct_mgmt },
	{ "open", "pam_open_session", pam_open_session },
	{ "close", "pam_close_session", pam_close_session },
	{ "pass", "pam_chauthtok", pam_chauthtok },
	{ NULL }
};

struct grouptab *
find_group(char *name)
{
	struct grouptab *p;
	for (p = grouptab; p->name; p++)
		if (strcmp(p->name, name) == 0)
			return p;
	return NULL;
}

void
groupprint()
{
	struct grouptab *p;
	for (p = grouptab; p->name; p++)
		printf("%s\n", p->name);
}

char *service = "check";
struct grouptab *group;
char *user;
char *pass;

static struct pam_conv conv = {
    pamck_conv,
    NULL
};

void
check_default(pam_handle_t *pamh)
{
	int rc;
	
        rc = pam_authenticate(pamh, 0);
	if (rc)
		error(2, "%s failed: %s",
		      "pam_authenticate", pam_strerror (pamh, rc));
	rc = pam_acct_mgmt(pamh, 0);
	if (rc)
		error(2, "%s failed: %s",
		      "pam_acct_mgmt", pam_strerror (pamh, rc));
	printf("OK\n");
}

void
check_group(pam_handle_t *pamh, struct grouptab *grp)
{
	int rc = grp->pam_fn(pamh, 0);
	if (rc) 
		error(2, "%s failed: %s", grp->funcname,
		      pam_strerror (pamh, rc));
}

int
main (int argc, char **argv)
{
	int c;
	int rc;
	pam_handle_t *pamh = NULL;
	
	program_name = argv[0];
	/* A bit of sugar to fake common GNU-style long options */
	if (argc == 2) {
		if (strcmp (argv[1], "--help") == 0) {
			help();
			exit(0);
		} if (strcmp (argv[1], "--usage") == 0) {
			usage();
			exit(0);
		} else if (strcmp (argv[1], "--version") == 0) {
			version();
			exit(0);
		}
	}
	/* Normal option processing */
	while ((c = getopt (argc, argv, "hg:s:v")) != EOF) {
		switch (c) {
		case 'h':
			help();
			exit(0);

		case 'g':
			if (strcmp(optarg, "help") == 0) {
				groupprint();
				exit(0);
			} 
			group = find_group(optarg);
			if (!group) 
				error(1,
				      "no such management group, try `%s -g help' for the list",
				      program_name);
				
			break;

		case 's':
			service = optarg;
			break;

		case 'v':
			version();
			exit(0);

		default:
			exit(1);
		}
	}

	argc -= optind;
	argv += optind;

	switch (argc) {
	case 2:
		pass = argv[1];
	case 1:
		user = argv[0];
		break;
	default:
		usage();
		exit(1);
	}

	rc = pam_start(service, user, &conv, &pamh);
	if (rc) 
		error(2, "pam_start failed");
	
	if (group)
		check_group(pamh, group);
	else
		check_default(pamh);

	if (pam_end(pamh, rc) != PAM_SUCCESS) {     
		pamh = NULL;
		error(2, "failed to release authenticator");
	}

	exit (0);
}
